types/src/rsmt2d.rs

```diff
@@ -159,45 +160,101 @@ impl ExtendedDataSquare {
+        let index = row * self.square_len + column;
+
+        self.data_square
+            .get(index)
+            .map(Vec::as_slice)
+            .ok_or(Error::EdsIndexOutOfRange(index))
+    }
+
+    /// Returns the shares of a row.
     pub fn row(&self, index: usize) -> Result<Vec<Vec<u8>>> {
-        Ok(self
-            .data_square
-            .get(index * self.square_len..(index + 1) * self.square_len)
-            .ok_or(Error::EdsIndexOutOfRange(index))?
-            .to_vec())
+        (0..self.square_len)
+            .map(|y| self.share(index, y).map(ToOwned::to_owned))
+            .collect()
+    }
+
+    /// Returns the [`Nmt`] of a row.
+    pub fn row_nmt(&self, index: usize) -> Result<Nmt> {
+        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
+
+        for y in 0..self.square_len {
+            let share = self.share(index, y)?;
+
+            let ns = if is_ods_square(index, y, self.square_len) {
+                Namespace::from_raw(&share[..NS_SIZE])?
+            } else {
+                Namespace::PARITY_SHARE
+            };
+
+            tree.push_leaf(share, *ns).map_err(Error::Nmt)?;
+        }
+
+        Ok(tree)
+    }
+
+    /// Returns the shares of a column.
+    pub fn column(&self, index: usize) -> Result<Vec<Vec<u8>>> {
+        (0..self.square_len)
+            .map(|x| self.share(x, index).map(ToOwned::to_owned))
+            .collect()
     }
 
-    /// Return colum with index
-    pub fn column(&self, mut index: usize) -> Result<Vec<Vec<u8>>> {
-        let mut r = Vec::with_capacity(self.square_len);
-        while index < self.data_square.len() {
-            r.push(
-                self.data_square
-                    .get(index)
-                    .ok_or(Error::EdsIndexOutOfRange(index))?
-                    .to_vec(),
-            );
-            index += self.square_len;
+    /// Returns the [`Nmt`] of a column.
+    pub fn column_nmt(&self, index: usize) -> Result<Nmt> {
+        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
+
+        for x in 0..self.square_len {
+            let share = self.share(x, index)?;
+
+            let ns = if is_ods_square(x, index, self.square_len) {
+                Namespace::from_raw(&share[..NS_SIZE])?
+            } else {
+                Namespace::PARITY_SHARE
+            };
+
+            tree.push_leaf(share, *ns).map_err(Error::Nmt)?;
         }
-        Ok(r)
+
+        Ok(tree)
     }
 
-    /// Return column or row with the provided index
+    /// Returns the shares of column or row.
     pub fn axis(&self, axis: AxisType, index: usize) -> Result<Vec<Vec<u8>>> {
         match axis {
             AxisType::Col => self.column(index),
             AxisType::Row => self.row(index),
         }
     }
 
-    /// Get EDS square length
+    /// Returns the [`Nmt`] of column or row.
+    pub fn axis_nmt(&self, axis: AxisType, index: usize) -> Result<Nmt> {
```

Comment:
`column_nmt` and `row_nmt` differ only when they fetch share from eds. I wonder whether I'd be more succinct to have shared code in `axis_nmt` and have `columnt_nmt` and `row_nmt` call it with appropriate `AxisType`?

Reply:
I was thinking about this too, but couldn't beside because indexes cause us a lot of confusion and I wanted to avoid a `match` that was selecting indexes. But I can change it.

Reply:
Not feeling too strongly about it, thought I'd make code more succint. I've had something like this in mind. 
```rust 
     pub fn axis_nmt(&self, axis: AxisType, index: usize) -> Result<Nmt> {
        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));

        let indices : Vec<_> = match axis {
            AxisType::Row => (0..self.square_len).zip(std::iter::repeat(index)).collect(),
            AxisType::Col => std::iter::repeat(index).zip(0..self.square_len).collect(),
        };

        for (x, y) in indices {
            let share = self.share(x, y)?;

            let ns = if is_ods_square(x, y, self.square_len) {
                Namespace::from_raw(&share[..NS_SIZE])?
            } else {
                Namespace::PARITY_SHARE
            };

            tree.push_leaf(share, *ns).map_err(Error::Nmt)?;
         }

        Ok(tree)
     }
```

Having to `collect` iterators, as they have different types, doesn't seem great though (also was this the match you were talking about? :P). So feel free to keep existing.

Reply:
No I was thinking for something like this:

```rust
    pub fn axis_nmt(&self, axis: AxisType, index: usize) -> Result<Nmt> {
        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));

        for i in 0..self.square_len {
            let (row, col) = match axis {
                AxisType::Row => (index, i),
                AxisType::Col => (i, index),
            };

            let share = self.share(row, col)?;

            let ns = if is_ods_square(col, row, self.square_len) {
                Namespace::from_raw(&share[..NS_SIZE])?
            } else {
                Namespace::PARITY_SHARE
            };

            tree.push_leaf(share, *ns).map_err(Error::Nmt)?;
        }

        Ok(tree)
    }
 ```

Reply:
done